File: apps/vivaah4you-web/src/lib/presence.ts
/**
 * Whether a member is around, from the `presence` block the API sends.
 *
 * The server decides what a viewer is allowed to know - an anonymous reader
 * gets no block at all, and for anyone but yourself the timestamp is floored to
 * the hour. This file only turns what arrived into something to render.
 */

/** "3 days ago" from an ISO timestamp, without pulling in a date library. */
export function relativeTime(iso: string | null): string {
  if (!iso) return ''
  const then = new Date(iso).getTime()
  if (Number.isNaN(then)) return ''

  const minutes = Math.round((Date.now() - then) / 60000)
  if (minutes < 1) return 'just now'
  if (minutes < 60) return `${minutes}m ago`
  const hours = Math.round(minutes / 60)
  if (hours < 24) return `${hours}h ago`
  const days = Math.round(hours / 24)
  if (days < 30) return `${days}d ago`
  return `${Math.round(days / 30)}mo ago`
}

/** The `presence` block as the API sends it. Absent for anonymous readers. */
export interface PresenceBlock {
  isOnline: boolean
  /** Null while they are online - "online now" is the whole message. */
  lastActiveAt: string | null
}

export interface Presence {
  online: boolean
  /** Empty when there is nothing honest to say - render nothing at all. */
  label: string
}

/**
 * A profile that has never been stamped returns an empty label rather than
 * "offline". Every profile predating the column is null, and calling those
 * members offline would be a claim the data does not support.
 */
export function presenceFor(presence: PresenceBlock | null | undefined): Presence {
  if (!presence) return { online: false, label: '' }
  if (presence.isOnline) return { online: true, label: 'Online now' }

  const ago = relativeTime(presence.lastActiveAt)
  return { online: false, label: ago ? `Last seen ${ago}` : '' }
}

File: services/vivaah4u-api/apps/profiles/presence.py
"""Whether a member is around right now.

Driven by a single `last_active_at` column stamped on authenticated API
activity - no sockets, no heartbeat endpoint, nothing to keep alive. The
tradeoff is resolution: someone reading a profile without calling anything
looks idle. That is acceptable for the one thing this drives, which is a dot
and a "last seen" line.
"""

from datetime import timedelta

from django.db import DatabaseError
from django.utils import timezone

from .models import Profile

#: Active within this window reads as "online now".
ONLINE_WINDOW = timedelta(minutes=5)

#: At most one write per member per interval, however many requests they make.
STAMP_INTERVAL = timedelta(minutes=1)


def touch(user) -> None:
    """Record that `user` is currently using the product.

    Three things here are load-bearing:

    - `.update()`, never `profile.save()`. `Profile.save()` recomputes
      completeness and the verification level on every call; running that on
      every API request would be absurd.
    - `.update()` also skips `auto_now`, so `updated_at` is left alone. That
      matters - `updated_at` is on the `/me` payload, and bumping it on every
      page load would invalidate every client's view of the profile.
    - The throttle is the WHERE clause rather than a read-then-write, so this
      is one conditional UPDATE with no SELECT, and it stays correct across
      worker processes.
    """
    if user is None:
        return

    now = timezone.now()
    try:
        Profile.objects.filter(user_id=user.id).exclude(
            last_active_at__gte=now - STAMP_INTERVAL
        ).update(last_active_at=now)
    except DatabaseError:
        # Presence is decoration. It must never turn a working request into a
        # 500 for the member trying to use the product.
        pass


def is_online(last_active_at) -> bool:
    return last_active_at is not None and timezone.now() - last_active_at <= ONLINE_WINDOW


def to_api(last_active_at, *, exact: bool) -> dict | None:
    """The presence block for a serialised profile, or None when unknown.

    `exact` is for a member looking at their own profile. Everywhere else the
    timestamp is floored to the hour: minute-resolution activity on someone
    else's profile is a usage log, and reconstructing one by polling is not
    something to leave open on a matrimonial product.

    A null column returns None rather than "offline" - every profile that
    predates this field has never been stamped, and calling them offline would
    be a claim we cannot support.
    """
    if last_active_at is None:
        return None

    online = is_online(last_active_at)
    stamp = last_active_at if exact else last_active_at.replace(minute=0, second=0, microsecond=0)

    return {
        "isOnline": online,
        # Online members report the window, not the instant - "online now" is
        # the whole message, and the exact minute adds nothing but exposure.
        "lastActiveAt": None if (online and not exact) else stamp.isoformat(),
    }
